adapter/authzOpa/authzOpa.go

```diff
@@ -0,0 +1,183 @@
+
+	rpc "github.com/googleapis/googleapis/google/rpc"
+	"github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/rego"
+
+	"istio.io/mixer/adapter/authzOpa/config"
+	"istio.io/mixer/pkg/adapter"
+	"istio.io/mixer/template/authz"
+)
+
+type (
+	builder struct {
+		adapterConfig *config.Params
+	}
+
+	handler struct {
+		policy      string
+		checkMethod string
+
+		context  context.Context
+		compiler *ast.Compiler
+
+		env adapter.Env
+	}
+)
+
+///////////////// Configuration Methods ///////////////
+
+func (b *builder) SetAuthzTypes(map[string]*authz.Type) {}
+
+func (b *builder) SetAdapterConfig(cfg adapter.Config) {
+	b.adapterConfig = cfg.(*config.Params)
+}
+
+func (b *builder) Validate() (ce *adapter.ConfigErrors) {
+	if len(b.adapterConfig.CheckMethod) == 0 {
+		ce = ce.Appendf(GetInfo().Name,
+			"CheckMethod was not configured")
+	}
+
+	parsed, err := ast.ParseModule("", string(b.adapterConfig.Policy))
+	if err != nil {
+		ce = ce.Appendf(GetInfo().Name,
+			"Failed to parse the OPA policy: %v", err)
+	}
+
+	compiler := ast.NewCompiler()
+	compiler.Compile(map[string]*ast.Module{"": parsed})
+	if compiler.Failed() {
+		ce = ce.Appendf(GetInfo().Name,
+			"Failed to compile the OPA policy: %v", compiler.Errors)
+	}
+
+	return nil
+}
+
+func (b *builder) Build(context context.Context, env adapter.Env) (adapter.Handler, error) {
+	h := handler{
+		policy:      b.adapterConfig.Policy,
+		checkMethod: b.adapterConfig.CheckMethod,
+		env:         env,
+	}
+
+	parsed, err := ast.ParseModule("", string(h.policy))
+	if err != nil {
+		return nil, errors.New(fmt.Sprintf(GetInfo().Name,
+			"Failed to parse the OPA policy: %v", err))
+	}
+
+	compiler := ast.NewCompiler()
+	compiler.Compile(map[string]*ast.Module{"": parsed})
+	if compiler.Failed() {
+		return nil, errors.New(fmt.Sprintf(GetInfo().Name,
+			"Failed to compile the OPA policy: %v", compiler.Errors))
+	}
+
+	h.compiler = compiler
+	h.context = context
+
+	return h, nil
+}
+
+////////////////// Runtime Methods //////////////////////////
+
+func (h handler) HandleAuthz(context context.Context, instance *authz.Instance) (adapter.CheckResult, error) {
+	variables := make(map[string]interface{})
+
+	for k, v := range instance.Subject {
+		variables[k] = v
+	}
+	for k, v := range instance.Resource {
+		variables[k] = v
+	}
+	for k, v := range instance.Verb {
+		variables[k] = v
+	}
+
+	rego := rego.New(
+		rego.Compiler(h.compiler),
+		rego.Query(h.checkMethod),
+		rego.Input(variables),
+	)
+
+	// Run evaluation.
+	rs, err := rego.Eval(h.context)
+
+	if err != nil {
+		return adapter.CheckResult{
+			Status: rpc.Status{Code: int32(rpc.PERMISSION_DENIED),
+				Message: fmt.Sprintf("authzOpa: request was rejected: %v", err)},
+		}, nil
+	}
+
+	if len(rs) != 1 {
+		return adapter.CheckResult{
+			Status: rpc.Status{Code: int32(rpc.PERMISSION_DENIED),
+				Message: "authzOpa: request was rejected"},
+		}, nil
+	}
+
+	result, ok := rs[0].Expressions[0].Value.(bool)
+	if !ok {
```

Comment:
Change this to `if !ok || result == false { ... }` and combine the two `ifs`.

Reply:
Applied